titan-compiler/checker.lua

```diff
@@ -689,43 +621,103 @@ local function checkfunc(node, st, errors)
-                    tlnode._type = modtype
-                    for _, err in ipairs(errs) do
-                        table.insert(errors, err)
-                    end
-                    st:add_symbol(tlnode.localname, tlnode)
-                else
-                    tlnode._type = types.Nil
-                    typeerror(errors, "problem loading module '%s': %s", tlnode._pos, name, errs)
+-- Return the name given the toplevel node
+local function toplevel_name(node)
+    local tag = node._tag
+    if tag == "TopLevel_Import" then
+        return node.localname
+    elseif tag == "TopLevel_Var" then
+        return node.decl.name
+    elseif tag == "TopLevel_Func" then
+        return node.name
+    else
+        error("tag not found " .. tag)
+    end
+end
+
+-- Typecheck the toplevel node
+local tlcheckers = {
+    ["TopLevel_Import"] =
+        function(node, st, errors, loader)
+            local modtype, errs = checker.checkimport(node.modname, loader)
+            if modtype then
+                node._type = modtype
+                for _, err in ipairs(errs) do
+                    table.insert(errors, err)
                 end
+            else
+                node._type = types.Nil
+                typeerror(errors, "problem loading module '%s': %s",
+                          node._pos, node.modname, errs)
             end
+        end,
+
+    ["TopLevel_Var"] =
+        function(node, st, errors)
+            if node.decl.type then
+                node._type = typefromnode(node.decl.type, errors)
+                checkexp(node.value, st, errors, node._type)
+                node.value = trycoerce(node.value, node._type, errors)
+                checkmatch("declaration of module variable " .. node.decl.name,
+                           node._type, node.value._type, errors, node._pos)
+            else
+                checkexp(node.value, st, errors)
+                node._type = node.value._type
+            end
+        end,
+
+    ["TopLevel_Func"] =
+        function(node, st, errors)
+            if #node.rettypes ~= 1 then
+                error("functions with 0 or 2+ return values are not yet implemented")
+            end
+            local ptypes = {}
+            for _, pdecl in ipairs(node.params) do
+                table.insert(ptypes, typefromnode(pdecl.type, errors))
+            end
+            local rettypes = {}
+            for _, rt in ipairs(node.rettypes) do
+                table.insert(rettypes, typefromnode(rt, errors))
+            end
+            node._type = types.Function(ptypes, rettypes)
+        end,
+}
+
+-- Colect type information of toplevel nodes
+--   ast: AST for the whole module
+--   st: symbol table
+--   errors: list of compile-time errors
+--   annotates the top-level nodes with their types in a "_type" field
+--   annotates whether a top-level declaration is duplicated with a "_ignore"
+--   field
+local function checktoplevel(ast, st, errors, loader)
```

Comment:
I am not sure if this is the exact same behavior as before. IIRC, the old version behaved like Javascript, effectively hoisting function variable declarations to the top of the module but keeping everything else the same while in this new version all the variable declarations are hoisted, as if there were a big letrec for the toplevel. Am I getting this right?

(In any case, I guess this special toplevel scoping is something we could discuss again)

Reply:
> I am not sure if this is the exact same behavior as before.

A regression test should be able clear this question!

Reply:
The older version sorted the declarations by: imports, variables, functions and then function bodies. This led to strange error messages:
```
function foo() end
foo = 0
```
This program had a duplicate declaration error in line 1, not line 2.
Is there any advantage in the old behavior that I'm not seeing?